doc/developer/guide-changes.md

```diff
@@ -0,0 +1,605 @@
+
+For example, the `sql-parser` and `sql` crates are entirely separate.
+The former crate tries very hard to limit its scope to *only* the lexical
+structure of the SQL language, without encoding any knowledge of the semantics
+of the SQL language. The parser knows that `SELECT * FROM bar` is well formed
+and `SELECT * FROM 'nope'` is not, but it has no idea what either of those
+statements mean.
+
+XXX still working out this paragraph. It's tempting to add logic to the
+`sql-parser` that rewrites SQL queries during parsing to make them easier to
+plan. It's a bit more work and code to add the appropriate hooks to the `sql`
+crate to transform queries as they come in, but it makes the SQL parser more
+understandable and more reusable, since it won't be rewriting queries behind
+your back in a way that folks besides the `sql` crate aren't interested in.
+
+### Dependencies
+
+One of the great dangers of the Rust ecosystem is the ease with which you can
+reach for a new dependency. Remember, every new dependency is code that we will
+indirectly have to maintain.
+
+Fixing a bug in a dependency is much harder than fixing a bug in core
+Materialize code. You'll have to check out the repository, spin up on the
+codebase, fix the bug, use Cargo's somewhat buggy [patch feature][cargo-patch]
+to verify the fix in Materialize, then send a PR to upstream. If upstream is
+responsive, then you just need to wait for a new version with the fix on
+crates.io. If upstream is not responsive, you'll need to fork the dependency,
+then keep that fork up-to-date if upstream ever becomes active again. Each step
+is a small papercut, but they add up.
+
+The best case is when a dependency is as actively maintained as Materialize.
+Dependencies that fall into this category include [Tokio], [rust-postgres],
+and [reqwest]. These dependencies have largely had no bugs at all. When they
+have had bugs, the upstream maintainers have been very quick to fix them.
+
+Use caution with dependencies that are not actively maintained. Materialize
+often winds up taking over maintainership of these dependencies. This can make
+sense for dependencies within our core competencies, like [rust-rdkafka]
+and [sqlparser], but it's a real drag for peripherals, like [rust-prometheus].
+
+### Documentation
+
+TODO @frankmcsherry?
+
+### Testing
+
+With perfect testing, introducing a bug in any line of any function in the
+codebase would cause at least one test to fail. Meeting this standard would
+unacceptably impair velocity, but it's a useful framing for what ideal test
+coverage means.
+
+The weaker property we strive for is that every change in behavior should
+require modifying or adding at _least_ one test. If a PR changes, say, the
+behavior of a SQL function and there are no changes to any test files, then
+there are certainly some missing tests.
+
+Our testing philosophy is covered in much greater detail in
+[Developer guide: testing](guide-testing.md).
+
```

Comment:
I would add a paragraph like the following:

Note that whenever possible, tests should validate _observable behavior_, not _implementation details_. Over-testing of trivial internal functions creates its own maintenance burden, as tests frequently have to be changed because of innocuous refactors, causing people to be less likely to attempt refactors, or worse, less likely to take test failures seriously.

If a piece of code is large and complex enough, it might warrant tests to validate its behavior, despite not being externally observable. Even in this case, tests should validate the behavior at a clean API boundary of this code, rather than its internal details. As always, use your best judgment.

Reply:
Thanks, done.